File: push/src/main/java/com/nihility/service/XMPushServiceListenerNotifier.java
package com.nihility.service;

import android.content.Intent;

import java.util.ArrayList;

public class XMPushServiceListenerNotifier implements XMPushServiceListener {
    private final ArrayList<XMPushServiceListener> listeners = new ArrayList<>();

    public void addListener(XMPushServiceListener listener) {
        listeners.add(listener);
    }

    @Override
    public void created() {
        for (XMPushServiceListener listener : listeners) {
            listener.created();
        }
    }

    @Override
    public void destroy() {
        for (XMPushServiceListener listener : listeners) {
            listener.destroy();
        }
    }

    @Override
    public void start(Intent intent) {
        for (XMPushServiceListener listener : listeners) {
            listener.start(intent);
        }
    }

    @Override
    public void connectionStatusChanged(ConnectionStatus connectionStatus) {
        for (XMPushServiceListener listener : listeners) {
            listener.connectionStatusChanged(connectionStatus);
        }
    }
}

File: xmsf/src/main/java/io/github/magisk317/mipush/service/XMPushServiceListenerNotifier.kt
package io.github.magisk317.mipush.service

import android.content.Intent
import io.github.magisk317.xposed.logging.MagiskOtel

open class XMPushServiceListenerNotifier : XMPushServiceListener {
    private val listeners = ArrayList<XMPushServiceListener>()

    fun addListener(listener: XMPushServiceListener) {
        listeners.add(listener)
    }

    override fun created() {
        for (listener in listeners) {
            listener.created()
        }
        emit(stage = "created", reason = "fanout", listenerCount = listeners.size)
    }

    override fun destroy() {
        for (listener in listeners) {
            listener.destroy()
        }
        emit(stage = "destroy", reason = "fanout", listenerCount = listeners.size)
    }

    override fun start(intent: Intent) {
        for (listener in listeners) {
            listener.start(intent)
        }
        emit(
            stage = "start",
            reason = intent.action?.take(64) ?: "no_action",
            listenerCount = listeners.size,
        )
    }

    override fun connectionStatusChanged(connectionStatus: ConnectionStatus) {
        for (listener in listeners) {
            listener.connectionStatusChanged(connectionStatus)
        }
        emit(
            stage = "connection_status",
            reason = connectionStatus.name.lowercase(),
            listenerCount = listeners.size,
        )
    }

    private fun emit(stage: String, reason: String, listenerCount: Int) {
        MagiskOtel.event(
            name = "push.lifecycle",
            attributes = mapOf(
                "result" to "ok",
                "duration_ms" to "0",
                "process" to "main",
                "stage" to stage,
                "reason" to reason,
                "pending_count" to listenerCount.toString(),
            ),
            statusOk = true,
        )
    }
}
